xrootd: add open request

EgorMatirov opened the pull request:
Updates go-hep/hep#170 (haven't included in the commit message since I expect rebases because of https://github.com/go-hep/hep/pull/214 and https://github.com/go-hep/hep/pull/211 and it will pollute issue with each commit)

Note that `open` request isn't added to `signing.go` because it creates a cycle import. Any suggestions on how to deal with that? Probably we can move `signing` to `protocol/signing` to fix that?

Also, note that there is a test workaround due to https://github.com/xrootd/xrootd/issues/721.

Also, I'm unsure about `UnmarshalXrd` in `EntryStat` and `NewEntryStat`. It duplicates logic, but otherwise  call to `UnmarshalXrd` in `dirlist.go`, `(Response) UnmarshalXrd(...)` will be like that:
```go
o.Entries[i/2].UnmarshalXrd(xrdenc.NewRBuffer([]byte(lines[i+1)]))
```

sbinet commented on xrootd/client/filesystem_test.go:
```diff
@@ -52,3 +53,128 @@ func TestFilesystem_Dirlist(t *testing.T) {
+		t.Errorf("Filesystem.Open()\ngotFileStat = %v\nwantFileStat = %v", gotFileStat, wantFileStat)
+	}
+
+	client.Close()
```
could we move it up to line 66 as a `defer client.Close()` ?

sbinet commented on xrootd/client/filesystem_test.go:
```diff
@@ -52,3 +53,128 @@ func TestFilesystem_Dirlist(t *testing.T) {
+		t.Errorf("Filesystem.Open()\ngotFileStat = %v\nwantFileStat = %v", gotFileStat, wantFileStat)
+	}
+
+	client.Close()
```
could we move it up to line 104 as a `defer`?

sbinet commented on xrootd/client/filesystem_test.go:
```diff
@@ -52,3 +53,128 @@ func TestFilesystem_Dirlist(t *testing.T) {
+		t.Errorf("Filesystem.Open()\ngotFileStat = %v\nwantFileStat = %v", gotFileStat, wantFileStat)
+	}
+
+	client.Close()
```
please move it up as a `defer` :)

sbinet commented on xrootd/protocol/open/open.go:
```diff
@@ -0,0 +1,157 @@
+	// OptionsOpenUpdate specifies that file is opened only for reading and writing.
+	OptionsOpenUpdate Options = 32
+	// OptionsAsyncOpen specifies that file is opened for asynchronous i/o.
+	OptionsAsync Options = 64
```
the doc says `OptionsAsyncOpen` but the `const` is called `OptionsAsync`.

sbinet commented on xrootd/protocol/protocol.go:
```diff
@@ -157,6 +158,91 @@ func (es EntryStat) Mode() os.FileMode {
+}
+
+// UnmarshalXrd implements xrootd/protocol.Unmarshaler
+func (o *EntryStat) UnmarshalXrd(rBuffer *xrdenc.RBuffer) error {
```
`dirlist` should probably be modified to reuse this code, right?
instead of writing it like:
```go
o.Entries[i/2].UnmarshalXrd(xrdenc.NewRBuffer([]byte(lines[i+1])))
```

I would do it like so:
```go
var rbuf =xrdenc.NewRBuffer(lines[i+1])
o.Entries[i/2].UnmarshalXrd(rbuf)
```
(we could have `dirlist` work in terms of `[][]byte` instead of converting to `[]string`.)

sbinet commented on xrootd/client/filesystem.go:
```diff
@@ -25,3 +26,14 @@ func (fs *FileSystem) Dirlist(ctx context.Context, path string) ([]protocol.Entr
 }
+
+// Open returns the file handle for a file together with the compression and the stat info.
+func (fs *FileSystem) Open(ctx context.Context, path string, mode open.Mode, options open.Options) (protocol.FileHandle, *protocol.FileCompression, *protocol.EntryStat, error) {
```
looking at the returned types, I have the urge to merge them into a single type, `File`, which would eventually implement `io.Reader`, `io.Writer`, `io.Closer`, ... with 2 additional methods:
- one to return this `FileCompression`
- the other to return `os.FileInfo` that our `EntryStat` implements already.

but perhaps it's a bit too soon to introduce such "high-level" abstractions? (or the wrong "level"?)

still... this really "feels" right:
```go
func (fs *FileSystem) Open(ctx context.Context, 
     path string, mode open.Mode, options open.Options) 
     (xrootd.File, error) { ... }
```

EgorMatirov replied:
I gave it a try. From my point of view, it looks fine, but I will be happy to hear your opinion. :)

sbinet commented:
as https://github.com/xrootd/xrootd/issues/721 has been fixed, could you add a blurb about that for the actions to take (if any) when a release with the fix percolates?
(also, if there are any actions to be taken on our side, please file an issue here so we don't forget :))

EgorMatirov commented:
Yep, I tweaked note a bit.

We haven't to do anything specific - just update the server after release and remove two lines (near that `TODO:` in the test) cause they will lose meaning. If you still want me to file the issue for that, then say and I'll be pleased to do that. :)

EgorMatirov commented:
Also, note that I have extracted an interface for `File` and moved file implementation to `Client` like it's done for `FileSystem`. It will help to implement a `Stat` request since it can operate on both the file path and the file handle.

So, there will be `Stat` in the `FileSystem` which returns `EntryStat` for the path and `FetchStat` in the `File` which will populate `File` with the fresh-fetched stat info by the file handle.

sbinet commented on xrootd/client/filesystem.go:
```diff
@@ -31,6 +32,17 @@ func (fs *fileSystem) Dirlist(ctx context.Context, path string) ([]xrdfs.EntrySt
+	if err != nil {
+		return file{}, err
+	}
+
```
please remove this blank line.

sbinet commented on xrootd/xrdfs/file.go:
```diff
@@ -0,0 +1,39 @@
+// File implements access to a content and meta information of file over XRootD.
+type File interface {
+	Compression() *FileCompression // Compression returns the compression info.
+	Info() *EntryStat              // Info returns the stat info.
```
same comment than on `client.file.Info`.
this should probably mention the possible staleness of the retrieved informations.

sbinet commented on xrootd/xrdproto/open/open.go:
```diff
@@ -0,0 +1,101 @@
+	Stat        *xrdfs.EntryStat
+}
+
+// MarshalXrd implements xrootd/protocol.Marshaler
```
`s|xrootd/protocol|xrdproto|`

sbinet commented on xrootd/xrdproto/open/open.go:
```diff
@@ -0,0 +1,101 @@
+	return nil
+}
+
+// UnmarshalXrd implements xrootd/protocol.Unmarshaler
```
`s|xrootd/protocol|xrdproto|`

sbinet commented on xrootd/xrdproto/open/open.go:
```diff
@@ -0,0 +1,101 @@
+	return nil
+}
+
+// RespID implements protocol.Response.RespID
```
`s/protocol/xrdproto/`

sbinet commented on xrootd/xrdproto/open/open.go:
```diff
@@ -0,0 +1,101 @@
+	return &Request{Mode: mode, Options: options, Path: path}
+}
+
+// MarshalXrd implements xrootd/protocol.Marshaler
```
`s|xrootd/protocol|xrdproto|`

sbinet commented on xrootd/xrdproto/open/open.go:
```diff
@@ -0,0 +1,101 @@
+	return nil
+}
+
+// UnmarshalXrd implements xrootd/protocol.Unmarshaler
```
`s|xrootd/protocol|xrdproto|`

sbinet commented on xrootd/xrdproto/open/open.go:
```diff
@@ -0,0 +1,101 @@
+	return nil
+}
+
+// ReqID implements protocol.Request.ReqID
```
`s|protocol|xrdproto|`

sbinet commented on xrootd/xrdproto/signing.go:
```diff
@@ -96,16 +97,20 @@ func NewSignRequirements(level SecurityLevel, overrides []SecurityOverride) Sign
 
 	if level >= Compatible {
 		// TODO: set requirements
+		sr.requirements[open.RequestID] = SignLikely
```
is the `TODO` still needed?

EgorMatirov replied:
Well, we need to add requirements for the most of requests, so I think we can leave `TODO` here to be sure.

sbinet replied:
ok.

sbinet commented on xrootd/xrdfs/file.go:
```diff
@@ -0,0 +1,39 @@
+	return nil
+}
+
+// UnmarshalXrd implements xrootd/protocol.Unmarshaler
```
`s|xrootd/protocol|xrdproto|`

sbinet commented on xrootd/xrdfs/file.go:
```diff
@@ -0,0 +1,39 @@
+	Type     [4]byte
+}
+
+// MarshalXrd implements xrootd/protocol.Marshaler
```
`s|xrootd/protocol|xrdproto|`

sbinet commented on xrootd/xrdfs/file.go:
```diff
@@ -0,0 +1,39 @@
+
+// FileCompression holds the compression parameters such as the page size and the type of compression.
+type FileCompression struct {
+	PageSize int32
```
it would be interesting to document in what unit (kB, MB, ...) this page size is.

EgorMatirov replied:
I was assuming that it's in bytes but it's unclear from the documentation (`cpsize` part): http://xrootd.org/doc/dev45/XRdv310.htm#_Toc464248823 :(

Same goes for the `cptype` - it's unclear from the documentation whether it's name or not.

I have tried to call `open` with `kxr_compress` option on `gzip`-ed file some time ago but haven't got any success with that.

sbinet replied:
feel free to send an issue asking for some clarifications of the specs :)

EgorMatirov replied:
I've filled an issue: https://github.com/xrootd/xrootd/issues/724.

sbinet replied:
thanks.

sbinet commented on xrootd/xrdfs/file.go:
```diff
@@ -0,0 +1,39 @@
+// FileCompression holds the compression parameters such as the page size and the type of compression.
+type FileCompression struct {
+	PageSize int32
+	Type     [4]byte
```
please document this exported field as well. (I guess it's the name of the compression algorithm?)

sbinet commented on xrootd/client/file.go:
```diff
@@ -0,0 +1,34 @@
+
+// File implements access to a content and meta information of file over XRootD.
+type file struct {
+	c           *Client
```
instead of `*Client`, wouldn't it make more sense to have `*fileSystem` instead?

EgorMatirov replied:
From some point of view, it is so, but I can't come up with the scenario when `fileSystem` will be needed in the `file` yet.

sbinet replied:
ok. well, one can always reach at the underlying `Client` from `*fileSystem` and get a `*fileSystem` from `Client` (although, in that case, at the price of a type-check.)

let's go with `fs *fileSystem` instead?

EgorMatirov commented on xrootd/client/file.go:
```diff
@@ -0,0 +1,34 @@
+	return f.compression
+}
+
+// Info returns the stat info.
```
You are right about the documentation.

Re `kXR_stat` I was thinking about something like:
```go
// FetchStatInfo updates the stat info of this file from the XRootD server.
func (f *file) FetchVirtualStatInfo(ctx context.Context) error {
	var resp stat.DefaultResponse
	err := f.c.Send(ctx, &resp, stat.NewRequestForHandle(f.handle))
	if err != nil {
		return err
	}
	f.info = &resp.EntryStat
	return nil
}
```
which will update this cached info.
This way user can easily re-use cached info and fetch fresh one (I'm not sure about naming though).

sbinet replied:
I think updating the cached `*info` field can be achieved as a (documented) side-effect of calling the yet-to-be-defined `file.Stat() (os.FileInfo, error)` (not sure yet whether we'd make it return `os.FileInfo` or `xrdfs.EntryStat`...)

sbinet commented:
(also, could you rebase?)

sbinet commented on xrootd/client/filesystem.go:
```diff
@@ -31,6 +32,16 @@ func (fs *fileSystem) Dirlist(ctx context.Context, path string) ([]xrdfs.EntrySt
+	if err != nil {
+		return file{}, err
+	}
+	return file{fs, resp.FileHandle, resp.Compression, resp.Stat}, nil
```
you probably want to `return &file{...}` here (and `return nil, err` above).
right now, both `client.file` and `*client.file` implement `xrdfs.File` but I suspect that later on, only `*client.file` will be able to implement it (_ie:_  for all the methods that modify the internal state of `client.file`, a pointer-receiver will be required.)

sbinet commented on xrootd/xrdfs/fs.go:
```diff
@@ -12,4 +12,61 @@ import (
+	// OpenOptionsNone specifies that file is opened without specific options.
+	OpenOptionsNone OpenOptions = 0
+	// OpenOptionsCompress specifies that file is opened even when compressed.
+	OpenOptionsCompress OpenOptions = 1
```
could we replace these with the explicit way of defining a bitmask? (ie: `1<<0`, `1<<1`, ... (or `1<<iota`))

EgorMatirov commented:
PTAL at https://github.com/go-hep/hep/pull/220 for the failure fix. :-/
